### PR TITLE
[SYCL][Matrix spec] keep deletion of assign op and copy ctor but change signature of joint_matrix_mad

### PR DESCRIPTION
**Pull request by dkhaldi**
@gmlueck, in https://github.com/intel/llvm/pull/7964, we explicitly deleted the copy ctor and assign op because we added `joint_matrix_copy` that is used to actually copy matrix data.
However, when deleting them in implementation, failures occur as `joint_matrix_mad` uses them.
This PR proposes to change the signature of `joint_matrix_mad` so these ctors are not used.

### CODE REVIEW

**Comment by dkhaldi** [comment]
gentle ping @gmlueck

File: sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc [hunk 1]
```diff
@@ -109,16 +109,17 @@ template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
...
 ```
-The constructor for the `joint_matrix` type is a group function as
-defined in section 4.17.3 of the core SYCL specification. It must be
-encountered in converged control flow by all work-items in the
-`Group`.
+The constructors for the `joint_matrix` type and the assignment
+operator are group functions as defined in section 4.17.3 of the core
+SYCL specification. They must be encountered in converged control flow
+by all work-items in the `Group`. Note that the assignment operator
+and the copy constructor do not copy the entire matrix content.
```
Comment (gmlueck) [inline sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:122]: Why do we need to say that the copy constructor and assignment operator don't copy the matrix content?  How would the application even be able to tell whether the matrix content was copied?  Isn't the content of the matrix hidden behind this abstract `joint_matrix` type?

I'm thinking now that we could replace `joint_matrix_copy` with an assignment operator that allows the user to copy from a matrix with different `Use` and `T` (type).  That way, matrix copy operations will always be done via the assignment operator.  That seems clearer than having both an assignment operator and a `joint_matrix_copy` function.

Comment (dkhaldi) [reply sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:122]: @gmlueck, if we have the assignment do the actual data copy, `= joint_matrix_mad` that uses assignment op would do extra unecessary copy of the data (https://github.com/intel/llvm/blob/faec54c77fdc706e013e5b6f2667592fba303ee4/sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp#L418)

Comment (gmlueck) [reply sycl/doc/extensions/experimental/sycl_ext_matrix/sycl_ext_oneapi_matrix.asciidoc:122]: It seems like neither `joint_matrix_copy` nor the assignment operator should copy anything to/from memory.  The only memory operations are `joint_matrix_load` and `joint_matrix_store`.  It seems like both `joint_matrix_copy` and the assignment operator should both just copy some "handle" that represents the register(s) that hold the matrix content?

In your implementation today, what is the difference between the implementation of `joint_matrix_copy` and the copy operator?

Here's another way to think about it.  Consider a code snippet like this:

```
joint_matrix<sub_group, int8_t, use::a, Rows, Cols, layout::row_major> m1;
joint_matrix<sub_group, int8_t, use::a, Rows, Cols, layout::row_major> m2;
joint_matrix_load(sg, m1, ptr1, Rows);  // Load "m1" with data from "ptr1"
m2 = m1;
joint_matrix_load(sg, m1, ptr2, Rows);  // Load "m1" with data from "ptr2"
```

I think a user would expect the matrix `m2` to still contain the data loaded from `ptr1` after the end of that code snippet.  Isn't this the same semantic as `joint_matrix_copy`?

**Comment by dkhaldi** [comment]
@gmlueck, can you please review?
This capture what we decided to do:
- keep deletion of copy ctor and assign op
- change signature of joint_matrix_mad by having D as input

I also did some cleanup: replace deprecated `multi_ptr()` with `address_space_cast`, replace one instance of `get_wi_data` (left by accident) by `joint_matrix_apply`.

**Review by gmlueck — COMMENTED** [review]
The changes look good, but the description of the PR is outdated:

> @gmlueck, in https://github.com/intel/llvm/pull/7964, we explicitly deleted the copy ctor and assign op because we added joint_matrix_copy that is used to actually copy matrix data.
However, when deleting them in implementation, failures occur as joint_matrix_mad uses them.
This PR proposes to keep them as implemented currently in https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp#L68
and add clarification that they don't copy matrix data.

Can you update the description, so the merged commit accurately describes what happened?

**Comment by dkhaldi** [comment]
> The changes look good, but the description of the PR is outdated:
> 
> > @gmlueck, in #7964, we explicitly deleted the copy ctor and assign op because we added joint_matrix_copy that is used to actually copy matrix data.
> > However, when deleting them in implementation, failures occur as joint_matrix_mad uses them.
> > This PR proposes to keep them as implemented currently in https://github.com/intel/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp#L68
> > and add clarification that they don't copy matrix data.
> 
> Can you update the description, so the merged commit accurately describes what happened?

Done

**Review by gmlueck — APPROVED** [review]

**Comment by dkhaldi** [comment]
@intel/llvm-gatekeepers, can you please merge this?

**Pull request merged**